ServerCore/Pages/Puzzles/Checklist.cshtml.cs

```diff
@@ -48,62 +48,48 @@ public async Task<IActionResult> OnGetAsync()
-                                                                 select new { Puzzle = prereqs.Key, Prereqs = (from p in prereqs orderby p.Prerequisite.Name select p.Prerequisite.Name).ToList() }).ToDictionaryAsync(x => x.Puzzle, x => x.Prereqs);
-            Dictionary<int, ResponseData> puzzleResponses = await (from response in _context.Responses
-                                                                   where response.Puzzle.Event == Event
-                                                                   group response by response.PuzzleID into responseList
-                                                                   select new { Puzzle = responseList.Key, Responses = new ResponseData { ResponseCount = responseList.Count(), HasAnswer = (from r in responseList where r.IsSolution select r).Count() > 0 } }).ToDictionaryAsync(x => x.Puzzle, x => x.Responses);
-            Dictionary<int, IEnumerable<int>> puzzleHints = await (from hint in _context.Hints
-                                                      where hint.Puzzle.Event == Event
-                                                      group hint by hint.Puzzle.ID into hints
-                                                      select new { Puzzle = hints.Key, Hints = (from h in hints select h.Cost) }).ToDictionaryAsync(x => x.Puzzle, x => x.Hints);
+            ILookup<int, string> puzzlePrereqs = (from prerequisite in _context.Prerequisites
+                                                  where prerequisite.Puzzle.Event == Event
+                                                  select prerequisite).ToLookup(prerequisite => prerequisite.PuzzleID, prerequisite => prerequisite.Prerequisite.Name);
+
+            HashSet<int> puzzlesWithSolutions = (from response in _context.Responses
+                                        where response.Puzzle.Event == Event && response.IsSolution
+                                        select response.PuzzleID).ToHashSet();
+            Dictionary<int, int> puzzleResponses = await (from response in _context.Responses
+                                                          where response.Puzzle.Event == Event
+                                                          group response by response.PuzzleID into responseList
+                                                          select new { Puzzle = responseList.Key, ResponseCount = responseList.Count() }).ToDictionaryAsync(x => x.Puzzle, x => x.ResponseCount);
+            var puzzleHints = await (from hint in _context.Hints
+                                     where hint.Puzzle.Event == Event
+                                     group hint by hint.Puzzle.ID into hints
+                                     select new { Puzzle = hints.Key, Count = hints.Count(), TotalDiscount = hints.Min(hint => hint.Cost), TotalCost = hints.Sum(hint => Math.Abs(hint.Cost)) }).ToDictionaryAsync(x => x.Puzzle);
```

Comment:
This looks like a functionality change for TotalCost, but I'd need to test some examples to be sure...

Reply:
You're probably right and I don't really understand this logic. Switched to a DB query that has the same result as the old one and restored the discount logic.